### ISSUE TITLE
@inline and deterministic results

### ISSUE THREAD

**Issue by lexaknyazev** [AssemblyScript/assemblyscript#2115]
Usually, the compiler makes its own assessment whether a function should be inlined at call-sites. However one could force it with an `@inline` decoration.

For some reason, adding the decoration to already-to-be-inlined functions sometimes lead to surprisingly different compilation output with two notable variations:
1. Increased number of local variables at call-site (although the callee is inlined anyway).
2. Inverted loop conditions, i.e., `for (let i: u32 = 0; i < c; i++)` the compiler puts the variables in reverse order and emits `i32.gt_u`.

Although these variations are quite minor, they sometimes decrease runtime performance and make debugging more difficult.

**Comment by MaxGraey** [comment]
I guess that's because inlining in binaryen comes with local optimizations usually ("inlining-optimization" pass). While inline on frontend side is always inline without local optimizations

**Comment by lexaknyazev** [comment]
Specifically, changing this:
```ts
namespace Bar {
    export function getFoo() : u32 {
        return load<u32>(CONST_LOCATION, 20);
    }
}
```

to this:
```ts
namespace Bar {
    @inline export function getFoo() : u32 {
        return load<u32>(CONST_LOCATION, 20);
    }
}
```
is enough to cause a regression in my case.

**Comment by dcodeIO** [comment]
Background here is that when the `@inline` decorator is present, the AssemblyScript compiler immediately inlines the function, no questions asked. If it is not present, Binaryen decides during optimization, based on optimization options, whether and where it makes sense to inline a function. So the difference manifests (only) when optimizing the module, with pass order affecting heuristics.

**Comment by lexaknyazev** [comment]
> potentially doing better than when doing so after the fact

That was my initial expectation. However in some cases binaryen's inline pass manages to produce a better output.

**Comment by MaxGraey** [comment]
Btw could you make minimal repro? I can't reproduce deterministic behaviour with your samples:
[playground](https://www.assemblyscript.org/editor.html#IyFvcHRpbWl6ZT1zcGVlZCZydW50aW1lPXN0dWIKY29uc3QgQ09OU1RfTE9DQVRJT04gPSA2NzsKCmV4cG9ydCBmdW5jdGlvbiBnZXRGb28oKSA6IHUzMiB7CiAgcmV0dXJuIGxvYWQ8dTMyPihDT05TVF9MT0NBVElPTiwgMjApOwp9CgpAaW5saW5lIGV4cG9ydCBmdW5jdGlvbiBnZXRGb29JbmxpbmVkKCkgOiB1MzIgewogIHJldHVybiBsb2FkPHUzMj4oQ09OU1RfTE9DQVRJT04sIDIwKTsKfQoKZXhwb3J0IGZ1bmN0aW9uIHRlc3QxKCk6IHUzMiB7CiAgcmV0dXJuIGdldEZvbygpICsgZ2V0Rm9vKCkgKyBnZXRGb28oKTsKfQoKZXhwb3J0IGZ1bmN0aW9uIHRlc3QyKCk6IHUzMiB7CiAgcmV0dXJuIGdldEZvb0lubGluZWQoKSArIGdldEZvb0lubGluZWQoKSArIGdldEZvb0lubGluZWQoKTsKfQojIWh0bWwKPHRleHRhcmVhIGlkPSJvdXRwdXQiIHN0eWxlPSJoZWlnaHQ6IDEwMCU7IHdpZHRoOiAxMDAlIiByZWFkb25seT48L3RleHRhcmVhPgo8c2NyaXB0Pgpsb2FkZXIuaW5zdGFudGlhdGUobW9kdWxlX3dhc20sIHsgLyogaW1wb3J0cyAqLyB9KQogIC50aGVuKCh7IGV4cG9ydHMgfSkgPT4gewogICAgY29uc3Qgb3V0cHV0ID0gZG9jdW1lbnQuZ2V0RWxlbWVudEJ5SWQoJ291dHB1dCcpCiAgICBmb3IgKGxldCBpID0gMDsgaSA8PSAxMDsgKytpKSB7CiAgICAgIG91dHB1dC52YWx1ZSArPSBgZmliKCR7aX0pID0gJHtleHBvcnRzLmZpYihpKX1cbmAKICAgIH0KICB9KQo8L3NjcmlwdD4=)

I am sure it is present, I would just like to understand under what conditions it occurs

**Comment by lexaknyazev** [comment]
Here's the smallest example I could get.

Config (`"converge": true` is required to repro):
```json
{
  "targets": {
    "release": {
      "optimizeLevel": 3,
      "shrinkLevel": 3,
      "converge": true,
      "noAssert": true
    }
  }
}
```

Source:
```ts
@inline function loadZ() : u32 {
  return load<u32>(16);
}

@inline function loadX(a: i32, b: i32): u32 {
  return load<u32>(2 * a + b);
}

export function baz(v: u32): void {
  if (v) unreachable();
}

export function foo(v0: i32, v1: i32, v2: i32): u32 {
  const x = loadX(v0, 0);
  const y = loadX(v0, 1);
  const z = loadZ();

  baz(z + x + y);

  struct.v0 = v0;
  struct.v1 = v1;
  struct.v2 = v2;

  return z;
}

@unmanaged class Struct {
  v0: u32;
  v1: u32;
  v2: u32;
}
const struct = changetype<Struct>(0);
```

With the very first `@inline` present (on `loadZ`), the `foo()` function has four local variables. Without it, only three.

**Comment by MaxGraey** [comment]
So `foo` function (without `@inline`):
```wat
(func $module/foo (param $0 i32) (param $1 i32) (param $2 i32) (result i32)
  (local $3 i32)
  local.get $0
  i32.const 1
  i32.shl
  local.tee $3
  i32.load offset=1
  local.get $3
  i32.load
  i32.const 16
  i32.load
  local.tee $3
  i32.add
  i32.add
  if
   unreachable
  end
  i32.const 0
  local.get $0
  i32.store
  i32.const 4
  local.get $1
  i32.store
  i32.const 8
  local.get $2
  i32.store
  local.get $3
 )
```

`foo` without `@inline` for `loadX` and `loadZ`:
```wat
(func $module/foo (param $0 i32) (param $1 i32) (param $2 i32) (result i32)
  (local $3 i32)
  (local $4 i32)
  local.get $0
  i32.const 1
  i32.shl
  local.tee $3
  i32.load
  local.set $4
  local.get $3
  i32.load offset=1
  i32.const 16
  i32.load
  local.tee $3
  local.get $4
  i32.add
  i32.add
  if
   unreachable
  end
  i32.const 0
  local.get $0
  i32.store
  i32.const 4
  local.get $1
  i32.store
  i32.const 8
  local.get $2
  i32.store
  local.get $3
 )
```

diff (with / without `@inline`s):
```diff
(func $module/foo (param $0 i32) (param $1 i32) (param $2 i32) (result i32)
  (local $3 i32)
+ (local $4 i32)
  local.get $0
  i32.const 1
  i32.shl
  local.tee $3
+ i32.load
+ local.set $4
+ local.get $3
  i32.load offset=1
- local.get $3
- i32.load
  i32.const 16
  i32.load
  local.tee $3
+ local.get $4
  i32.add
  i32.add
  if
   unreachable
  end
  i32.const 0
  local.get $0
  i32.store
  i32.const 4
  local.get $1
  i32.store
  i32.const 8
  local.get $2
  i32.store
  local.get $3
 )

**Comment by MaxGraey** [comment]
this could be fixed by adding early `inline` pass #2117
